gr-blocks/lib/correctiq_auto_impl.cc

```diff
@@ -0,0 +1,123 @@
+correctiq_auto::sptr
+correctiq_auto::make(float samp_rate, float freq, float gain, float syncWindow)
+{
+    return gnuradio::get_initial_sptr(
+        new correctiq_auto_impl(samp_rate, freq, gain, syncWindow));
+}
+
+/*
+ * The private constructor
+ */
+correctiq_auto_impl::correctiq_auto_impl(float samp_rate,
+                                         float freq,
+                                         float gain,
+                                         float syncWindow)
+    : gr::sync_block("correctiq_auto",
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)),
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)))
+{
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    d_samp_rate = samp_rate;
+    d_freq = freq;
+    d_gain = gain;
+    d_syncWindow = syncWindow;
+
+    synchronized = false;
+
+    std::cout << "CorrectIQ Auto Synchronizing..." << std::endl;
+
+    d_max_syncSamples = (long)(d_samp_rate * syncWindow);
+}
+
+/*
+ * Our virtual destructor.
+ */
+correctiq_auto_impl::~correctiq_auto_impl() {}
+
+float correctiq_auto_impl::get_freq() { return d_freq; }
+
+float correctiq_auto_impl::get_gain() { return d_gain; }
+
+void correctiq_auto_impl::set_freq(float new_value)
+{
+    std::cout << "[CorrectIQ_Auto] Auto Synchronizing..." << std::endl;
+    d_freq = new_value;
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+}
+
+void correctiq_auto_impl::set_gain(float new_value)
+{
+    std::cout << "[CorrectIQ_Auto] Auto Synchronizing..." << std::endl;
+    d_gain = new_value;
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+}
+
+int correctiq_auto_impl::work(int noutput_items,
+                              gr_vector_const_void_star& input_items,
+                              gr_vector_void_star& output_items)
+{
+    const gr_complex* in = (const gr_complex*)input_items[0];
+    gr_complex* out = (gr_complex*)output_items[0];
+
+    for (int i = 0; i < noutput_items; i++) {
+        if (synchronized) {
+            out[i].real(in[i].real() - avg_real);
+            out[i].imag(in[i].imag() - avg_img);
+        } else {
+            // Synchronizing.  Behave just like normal correctiq.
+            avg_real = ratio * (in[i].real() - avg_real) + avg_real;
+            avg_img = ratio * (in[i].imag() - avg_img) + avg_img;
+
+            out[i].real(in[i].real() - avg_real);
+            out[i].imag(in[i].imag() - avg_img);
+
+            sync_counter++;
+        }
+    }
+
+    if (!synchronized && (sync_counter >= d_max_syncSamples)) {
+        synchronized = true;
+
+        std::cout << "CorrectIQ Auto offset now synchronized." << std::endl;
+        std::cout << "Applying these offsets... Real:" << std::fixed << std::setw(11)
+                  << std::setprecision(6) << -avg_real << ", imag:" << -avg_img
+                  << std::endl;
```

Comment:
when logging formatted strings, you can either use `std::stringstream` to generate a string, or, very popular, `boost::format()`, which might be a bit easier to use. 

I really like that this generates a lot of debug info – but that really illustrates how important it is to keep cout clean in case someone's looking for something else in there.